plugins/secretstores/http/aes.go

```diff
@@ -0,0 +1,174 @@
+		a.trim = func(in []byte) ([]byte, error) { return in, nil }
+	case "pkcs5", "pkcs#5", "pkcs5padding", "pkcs#5padding":
+		// The implementation can handle both variants, so fallthrough to
+		// the PKCS#7 case
+		fallthrough
+	case "pkcs7", "pkcs#7", "pkcs7padding", "pkcs#7padding":
+		a.trim = PKCS7Trimming
+	default:
+		return fmt.Errorf("unsupported padding %q", padding)
+	}
+
+	// Generate the key using password-based-keys
+	if a.Key.Empty() {
+		if a.Passwd.Empty() {
+			return errors.New("either key or password has to be specified")
+		}
+		if a.Salt.Empty() || a.Iterations == 0 {
+			return errors.New("salt and iterations required for password-based-keys")
+		}
+
+		key, iv, err := a.KDFConfig.NewKey(keylen)
+		if err != nil {
+			return fmt.Errorf("generating key failed: %w", err)
+		}
+		if err := a.Key.Set([]byte(hex.EncodeToString(key))); err != nil {
+			return fmt.Errorf("setting key failed: %w", err)
+		}
+
+		if a.Vec.Empty() && len(iv) > 0 {
+			if err := a.Vec.Set(iv); err != nil {
+				return fmt.Errorf("setting IV failed: %w", err)
+			}
+		}
+	} else {
+		encodedKey, err := a.Key.Get()
+		if err != nil {
+			return fmt.Errorf("getting key failed: %w", err)
+		}
+		key := make([]byte, hex.DecodedLen(len(encodedKey)))
+		if _, err := hex.Decode(key, encodedKey); err != nil {
+			config.ReleaseSecret(encodedKey)
+			return fmt.Errorf("decoding key failed: %w", err)
+		}
+		config.ReleaseSecret(encodedKey)
+
+		if len(key) != keylen {
+			return fmt.Errorf("key length (%d bit) does not match cipher (%d bit)", len(key)*8, keylen*8)
+		}
+	}
+
+	if a.Vec.Empty() {
+		return errors.New("'init_vector' has to be specified or derived from password")
+	}
+
+	return nil
+}
+
+func (a *AesEncryptor) Decrypt(data []byte) ([]byte, error) {
+	if len(data)%aes.BlockSize != 0 {
+		return nil, fmt.Errorf("invalid data size %d", len(data))
+	}
+
+	// Setup the cipher and return the decoded data
+	encodedKey, err := a.Key.Get()
+	if err != nil {
+		return nil, fmt.Errorf("getting key failed: %w", err)
+	}
+	key := make([]byte, hex.DecodedLen(len(encodedKey)))
+	if _, err := hex.Decode(key, encodedKey); err != nil {
+		config.ReleaseSecret(encodedKey)
+		return nil, fmt.Errorf("decoding key failed: %w", err)
+	}
+	config.ReleaseSecret(encodedKey)
+
+	// Setup AES
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, fmt.Errorf("creating AES cipher failed: %w", err)
+	}
+
+	// Setup the block/stream cipher and decode the data
+	iv, err := a.Vec.Get()
+	if err != nil {
+		return nil, fmt.Errorf("getting initialization-vector failed: %w", err)
+	}
+	defer config.ReleaseSecret(iv)
+
+	switch a.mode {
+	case "none":
+		block.Decrypt(data, data)
+	case "cbc":
+		cipher.NewCBCDecrypter(block, iv).CryptBlocks(data, data)
+	case "cfb":
+		cipher.NewCFBDecrypter(block, iv).XORKeyStream(data, data)
+	case "ctr":
+		cipher.NewCTR(block, iv).XORKeyStream(data, data)
+	case "ofb":
+		cipher.NewOFB(block, iv).XORKeyStream(data, data)
```

Comment:
All these `...(data, data)` calls modifies `data` as a side-effect of this function, no? This doesn't seem to be an issue per se with the current implementation and how `PKCS7Trimming` works, but it seems like it could be if the implementation changed.

Reply:
The [documentation](https://pkg.go.dev/crypto/cipher#Stream) says:
> 	XORKeyStream XORs each byte in the given slice with a byte from the
>	cipher's key stream. Dst and src must overlap entirely or not at all.

so to me this is a guarantee this interface makes and it is inherent to XOR or trimming that the above works. So I would rather avoid doing data copying here but if you want me to, I will change the code!?!?

Reply:
No, that's fine. Feel free to resolve.